internals/review_models.py

```diff
@@ -188,3 +188,131 @@ def get_raw_owner_file(cls, url):
+  next_action = ndb.DateProperty()
+  additional_review = ndb.BooleanProperty(default=False)
+
+  # TODO(jrobbins): implement request_review() and clear_request()
+
+
+# Note: This class is not used yet.
+class Vote(ndb.Model):  # copy from Approval
+  """One approver's vote on what the state of a gate should be."""
+
+  # Not used: NEEDS_REVIEW = 0
+  NA = 1
+  REVIEW_REQUESTED = 2
+  REVIEW_STARTED = 3
+  NEED_INFO = 4
+  APPROVED = 5
+  NOT_APPROVED = 6
+  VOTE_VALUES = {
+      # Not used: NEEDS_REVIEW: 'needs_review',
+      NA: 'na',
+      REVIEW_REQUESTED: 'review_requested',
+      REVIEW_STARTED: 'review_started',
+      NEED_INFO: 'need_info',
+      APPROVED: 'approved',
+      NOT_APPROVED: 'not_approved',
+  }
+
+  FINAL_STATES = [NA, APPROVED, NOT_APPROVED]
+
+  feature_id = ndb.IntegerProperty(required=True)
+  gate_id = ndb.IntegerProperty(required=True)
+  state = ndb.IntegerProperty(required=True)
+  set_on = ndb.DateTimeProperty(required=True)
+  set_by = ndb.StringProperty(required=True)
+
+  @classmethod
+  def get_votes(
+      cls, feature_id=None, gate_id=None, states=None, set_by=None,
+      limit=None):
+    """Return the requested approvals."""
+    query = Vote.query().order(Approval.set_on)
+    if feature_id is not None:
+      query = query.filter(Vote.feature_id == feature_id)
+    if gate_id is not None:
+      query = query.filter(Vote.gate_id == gate_id)
+    if states is not None:
+      query = query.filter(Vote.state.IN(states))
+    if set_by is not None:
+      query = query.filter(Vote.set_by == set_by)
+    # Query with STRONG consistency because ndb defaults to
+    # EVENTUAL consistency and we run this query immediately after
+    # saving the user's change that we want included in the query.
+    votes = query.fetch(limit, read_consistency=ndb.STRONG)
+    return votes
+
+  @classmethod
+  def is_valid_state(cls, new_state):
+    """Return true if new_state is valid."""
+    return new_state in cls.VOTE_VALUES
+
+  @classmethod
+  def set_vote(cls, feature_id, gate_id, new_state, set_by_email):
+    """Add or update an approval value."""
+    if not cls.is_valid_state(new_state):
+      raise ValueError('Invalid approval state')
+
+    now = datetime.datetime.now()
+    existing_list = cls.get_votes(
+        feature_id=feature_id, gate_id=gate_id, set_by=set_by_email)
+    if existing_list:
+      existing = existing_list[0]
+      existing.set_on = now
+      existing.state = new_state
+      existing.put()
+      return
+
+    new_vote = Vote(
+        feature_id=feature_id, gate_id=gate_id, state=new_state,
+        set_on=now, set_by=set_by_email)
+    new_vote.put()
+
+  # Note: clear_request() moved to Gate
+
+
+# Note: This class is not used yet.
+class Amendment(ndb.Model):
+  """ReviewComments can log changes to other fields of the feature."""
+  field_name = ndb.StringProperty()  # from QUERIABLE_FIELDS
+  old_value = ndb.TextProperty()
+  new_value = ndb.TextProperty()
```

Comment:
Is there any merit in noting the user who made an amendment?

Reply:
Will the author always be the one making an amendment? I would think so, and if not, it would be a site admin. I imagine the case is so rare that it's just not worth keeping track of. Thanks for answering my question!

Reply:
I think we're going to want to know who made a particular change auditing purposes. If I read this correctly, I can tell the changes made by a particular author, but I can't tell the author of a particular change.

Reply:
I'm thinking that any time that a user edits a feature, the app will generate a set of Amendments just like the way that we generate diffs for the notifications now.  Rather than being used only for email, those Amendments would also be put into a ReviewComment.  That ReviewComment would have the author set to the email address of the person who made the change, and the content of the comment would be empty.

The ID of the ReviewComment entity could be passed to notifier.py to generate the notification email.  And, that same ReviewComment could be displayed somewhere on the feature detail page (regardless of gate ID), and in the approvals dialog box (only for the relevant gate ID).  

I'll write that up as an issue.